### PR TITLE
[cpu] add sdpa choice and UT

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105131
* #107128
* #104863
* #104693
* #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

Write an SDPA selecting function for CPU to automatically choose one SDPA implementation among several ones. There are two CPU implementations which could be chosen: the unfused SDPA and flash attention. In general, flash attention has a higher priority than the unfused SDPA. For cases where flash attention is not applicable, such as manually disabling flash attention or the inputs not 4 dimensional, the unfused SDPA is chosen.

### NanoGPT's SDPA kernel
Using benchmark [repo](https://github.com/mingfeima/bench_sdpa/blob/main/README.md), with one socket.
Shape: Batch size 1, Sequence length 1024, Head number 25, Head size 64.
Machine: SPR.

| Dtype    | Causal   | Mode      | SDPA            | Time (ms per iter) | Speedup |
| -------- | -------- | -------   | -------         | -------            | ------- |
| float32  | FALSE    | Inference | Unfused         | 3.081              |         |
|          |          |           | Flash attention | 1.665              | **1.85045** |
| float32  | TRUE     | Inference | Unfused         | 3.463              |         |
|          |          |           | Flash attention | 1.662              | **2.083634**|
| bfloat16 | FALSE    | Inference | Unfused         | 1.203              |         |
|          |          |           | Flash attention | 1.154              | **1.042461**|
| bfloat16 | TRUE     | Inference | Unfused         | 1.543              |         |
|          |          |           | Flash attention | 1.154              | **1.337088**|
| float32  | FALSE    | Training  | Unfused         | 54.938             |         |
|          |          |           | Flash attention | 23.029             | **2.385601**|
| float32  | TRUE     | Training  | Unfused         | 58.266             |         |
|          |          |           | Flash attention | 17.835             | **3.266947**|
| bfloat16 | FALSE    | Training  | Unfused         | 18.924             |         |
|          |          |           | Flash attention | 18.886             | **1.002012**|
| bfloat16 | TRUE     | Training  | Unfused         | 21.08              |         |
|          |          |           | Flash attention | 14.172             | **1.48744** |

### Stable Diffusion
Following model's [BKM](https://github.com/intel-innersource/frameworks.ai.models.intel-models/blob/develop/quickstart/diffusion/pytorch/stable_diffusion/inference/cpu/README.md).
Mode: Inference; Machine: SPR.

| Dtype    | SDPA                    | Throughput (fps) | Speedup SDPA | Total Time (ms) | Speedup |
| -------- | --------                | -------          | -------      | -------         | ------- |
| float32  | Unfused                 | 1.63             |              | 1139            |         |
|          | Flash attention         | 1.983            | 1.216564     | 547.488         | **2.080411**|
| bfloat16 | Flash attention in IPEX | 4.784            |              | 429.051         |         |
|          | Flash attention         | 4.857            | 1.015259     | 408.823         | **1.049479**|

### LLM models of Torchbench

Dtype: float32; Mode: Inference, single socket; Machine: CPX.
Model   name | SDPA | Inductor_new | Inductor_old | Inductor   Ratio(old/new)
-- | -- | -- | -- | --
hf_Albert | Unfused -> Flash attention | 0.048629309 | 0.05591545 | **1.14983024**
hf_Bert | Unfused -> Flash attention | 0.053156243 | 0.060732115 | **1.142520841**
hf_Bert_large | Unfused -> Flash attention | 0.141089502 | 0.155190077 | **1.099940636**
llama | Unfused -> Flash attention | 0.033250106 | 0.033720745 | **1.01415451**

Dtype: bfloat16; Mode: Inference, single socket; Machine: SPR.
Model   name | SDPA | Inductor_new | Inductor_old | Inductor   Ratio(old/new)
-- | -- | -- | -- | --
hf_Albert | Unfused -> Flash attention | 0.020681298 | 0.020718282 | **1.001788324**
hf_Bert | Unfused -> Flash attention | 0.019932816 | 0.019935424 | **1.000130842**
hf_Bert_large | Unfused -> Flash attention | 0.047949174 | 0.048312502 | **1.007577355**
llama | Unfused -> Flash attention | 0.018528057 | 0.01861126 | **1.0044907**

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp:793]: Thanks and updated.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp:805]: It is found that some test cases could run into `cpu_flash_attention_backward` with `is_training=False`, so I keep the calculation of `logsumexp` at present.

Comment (jgong5) [reply aten/src/ATen/native/transformers/attention.cpp:807]: Is it a test case issue?

Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp:807]: Yes, it looks like a test case issue. The issue can be reproduced by `python test/test_ops.py -k test_backward_nn_functional_scaled_dot_product_attention_cpu_float32`. The test case would change a subset of inputs from `Tensor` to `CompositeCompliantTensor`. By doing this, `input.requires_grad()` would change from `true` to `false`.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:798]: Curious why this isn't identically equal to 0.0

File: aten/src/ATen/native/transformers/attention.cpp
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:794]: Also, ideally these torch_checks should never get hit since we sdp_utils_cpp dispatcher knows that flash cpu implementation can not be run if any of these are set.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (drisspg) [reply aten/src/ATen/native/transformers/attention.cpp:807]: Yeah this is a test case issue, its probs better to keep the is_training check and skip the particular failing test

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:1511]: we should just move this into a parametrized pytest value

File: aten/src/ATen/native/transformers/sdp_utils_cpp.cpp
Comment (drisspg) [inline aten/src/ATen/native/transformers/sdp_utils_cpp.cpp:86]: If we aren't going to implement this we should just remove

Comment (drisspg) [reply aten/src/ATen/native/transformers/sdp_utils_cpp.cpp:86]: And all the commented out parts as well

File: torch/testing/_internal/common_methods_invocations.py
Comment (drisspg) [inline torch/testing/_internal/common_methods_invocations.py:13471]: can you elaborate on this, and would cause any issue when registering the correct meta function?

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:1550]: Is it possible to use some of the same helpers for creating inputs defined here: 
https://github.com/pytorch/pytorch/blob/b08b0c915f59cce79630a027fe501f54b6b15e4f/test/test_transformers.py#L101

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp]: Sure! I have added the `is_training` back and skipped the UT.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp]: Modified to `0.0`.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp:798]: Absolutely. However, it may be hit if someone directly calls `torch.ops.aten._scaled_dot_product_flash_attention`. So, I still keep the `torch_checks` here, just in case.

File: aten/src/ATen/native/transformers/sdp_utils_cpp.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/sdp_utils_cpp.cpp]: Thanks and modified.

**Comment by Valentine233** [comment]
@zou3519 @Chillee @kshitij12345 @mruberry @ngimel Hi, please help review the PR, as this needs the approval from code owners.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (drisspg) [reply aten/src/ATen/native/transformers/attention.cpp:794]: Yup makes sense just making sure

**Review by drisspg — APPROVED** [review]

**Pull request closed without merging**
